sdk/nodejs/output.ts

```diff
@@ -312,6 +236,77 @@ To manipulate the value of this Output, use '.apply' instead.`);
             },
         });
     }
+
+    public get(): T {
+        throw new Error(`Cannot call '.get' during update or preview.
+To manipulate the value of this Output, use '.apply' instead.`);
+    }
+
+    public apply<U>(func: (t: T) => Input<U>): Output<U> {
+        const applied: Promise<OutputData<U>> =
+            Promise.all([this.promise(), this.isKnown, this.isSecret])
+                   .then(([value, isKnown, isSecret]) => applyHelperAsync<T, U>(value, isKnown, isSecret, func));
+
+        const result = new OutputImpl<U>(
+            this.resources(),
+            applied.then(a => a.value),
+            applied.then(a => a.isKnown),
+            applied.then(a => a.isSecret));
+        return <Output<U>><any>result;
+    }
+}
+
+interface OutputData<T> {
+    value: T;
+    isKnown: boolean;
+    isSecret: boolean;
+}
+
+async function applyHelperAsync<T, U>(value: T, isKnown: boolean, isSecret: boolean, func: (t: T) => Input<U>): Promise<OutputData<U>> {
+    if (runtime.isDryRun()) {
+        // During previews only perform the apply if the engine was able to
+        // give us an actual value for this Output.
+        const applyDuringPreview = isKnown;
+
+        if (!applyDuringPreview) {
+            // We didn't actually run the function, our new Output is definitely
+            // **not** known.
+            return {
+                value: <U><any>undefined,
+                isKnown: false,
+                isSecret,
```

Comment:
Preserving the behavior from before: https://github.com/pulumi/pulumi/pull/3353/files#diff-b7a9c3c1793336a377df01cb5e0e9f40L202.  During preview, even if we don't apply the func, we preserve the secret-bit.